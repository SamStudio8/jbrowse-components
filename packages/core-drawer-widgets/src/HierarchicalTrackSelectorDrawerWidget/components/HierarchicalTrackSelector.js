import Fab from '@material-ui/core/Fab'
import Icon from '@material-ui/core/Icon'
import IconButton from '@material-ui/core/IconButton'
import InputAdornment from '@material-ui/core/InputAdornment'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import Paper from '@material-ui/core/Paper'
import Tab from '@material-ui/core/Tab'
import Tabs from '@material-ui/core/Tabs'
import { withStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import { PropTypes as MobxPropTypes } from 'mobx-react'
import { observer } from 'mobx-react-lite'
import { getRoot } from 'mobx-state-tree'
import propTypes from 'prop-types'
import React, { useState } from 'react'
import { readConfObject } from '@gmod/jbrowse-core/configuration'
import Contents from './Contents'

const styles = theme => ({
  root: {
    textAlign: 'left',
    padding: theme.spacing.unit,
  },
  searchBox: {
    marginBottom: theme.spacing.unit * 2,
  },
  fab: {
    float: 'right',
    position: 'sticky',
    marginTop: theme.spacing.unit * 2,
    bottom: theme.spacing.unit * 2,
    right: theme.spacing.unit * 2,
  },
  connectionsPaper: {
    padding: theme.spacing.unit,
    marginTop: theme.spacing.unit,
  },
  tabs: {
    marginBottom: theme.spacing.unit,
  },
})

function HierarchicalTrackSelector(props) {
  const [anchorEl, setAnchorEl] = useState(null)
  const [assemblyIdx, setAssemblyIdx] = useState(0)

  const { model, classes } = props
  const rootModel = getRoot(model)

  function handleTabChange(event, newIdx) {
    setAssemblyIdx(newIdx)
  }

  function handleFabClick(event) {
    setAnchorEl(event.currentTarget)
  }

  function handleFabClose() {
    setAnchorEl(null)
  }

  function handleInputChange(event) {
    model.setFilterText(event.target.value)
  }

  function addDataHub() {
    handleFabClose()
    if (!rootModel.drawerWidgets.get('dataHubDrawerWidget'))
      rootModel.addDrawerWidget('DataHubDrawerWidget', 'dataHubDrawerWidget')
    rootModel.showDrawerWidget(
      rootModel.drawerWidgets.get('dataHubDrawerWidget'),
    )
  }

  function addTrack() {
    handleFabClose()
    rootModel.addDrawerWidget('AddTrackDrawerWidget', 'addTrackDrawerWidget', {
      view: model.view.id,
    })
    rootModel.showDrawerWidget(
      rootModel.drawerWidgets.get('addTrackDrawerWidget'),
    )
  }

  function filter(trackConfig) {
    if (!model.filterText) return true
    const name = readConfObject(trackConfig, 'name')
    return name.toLowerCase().includes(model.filterText.toLowerCase())
  }

  const { assemblyNames } = model
  const assemblyName = assemblyNames[assemblyIdx]
  const filterError =
    model.trackConfigurations(assemblyName).filter(filter).length === 0

  return (
    <div key={model.view.id} className={classes.root}>
      {assemblyNames.length > 1 ? (
        <Tabs
          className={classes.tabs}
          value={assemblyIdx}
          onChange={handleTabChange}
        >
          {assemblyNames.map(name => (
            <Tab key={name} label={name} />
          ))}
        </Tabs>
      ) : null}
      <TextField
        className={classes.searchBox}
        label="Filter Tracks"
        value={model.filterText}
        error={filterError}
        helperText={filterError ? 'No matches' : ''}
        onChange={handleInputChange}
        fullWidth
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">
              <IconButton onClick={model.clearFilterText}>
                <Icon>clear</Icon>
              </IconButton>
            </InputAdornment>
          ),
        }}
      />
      <Contents
        model={model}
        filterPredicate={filter}
        assemblyName={assemblyName}
        top
      />
      {rootModel.connections.size ? (
        <>
          <Typography variant="h5">Connections:</Typography>
          {Array.from(rootModel.connections.keys()).map(connectionName => (
            <Paper
              key={connectionName}
              className={classes.connectionsPaper}
              elevation={8}
            >
              <Typography variant="h6">{connectionName}</Typography>
              <Contents
                model={model}
                filterPredicate={filter}
                connection={connectionName}
                assemblyName={assemblyName}
                top
              />
            </Paper>
          ))}
        </>
      ) : null}

      <Fab color="secondary" className={classes.fab} onClick={handleFabClick}>
        <Icon>add</Icon>
      </Fab>
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        open={Boolean(anchorEl)}
        onClose={handleFabClose}
      >
        <MenuItem onClick={addDataHub}>Add Data Hub</MenuItem>
        <MenuItem onClick={addTrack}>Add track</MenuItem>
      </Menu>
    </div>
  )
}

HierarchicalTrackSelector.propTypes = {
  classes: propTypes.objectOf(propTypes.string).isRequired,
  model: MobxPropTypes.observableObject.isRequired,
}

export default withStyles(styles)(observer(HierarchicalTrackSelector))
