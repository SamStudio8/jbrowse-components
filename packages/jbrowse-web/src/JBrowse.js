import '@gmod/jbrowse-core/fonts/material-icons.css'
import CircularProgress from '@material-ui/core/CircularProgress'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/styles'
import { observer } from 'mobx-react'
import React from 'react'
import 'typeface-roboto'
import App from './ui/App'
import Theme from './ui/theme'

export default observer(({ state }) => {
  let DisplayComponent = (
    <CircularProgress
      style={{
        position: 'fixed',
        top: '50%',
        left: '50%',
        marginTop: -250,
        marginLeft: -250,
      }}
      size={500}
    />
  )
  if (state.errorMessage) DisplayComponent = <div>{state.errorMessage}</div>
  if (state.activeSession)
    DisplayComponent = (
      <App
        session={state.activeSession}
        sessionNames={state.sessionNames}
        setActiveSession={state.activateSession}
        addSession={state.addSession}
      />
    )

  return (
    <ThemeProvider theme={Theme}>
      <CssBaseline />
      {DisplayComponent}
    </ThemeProvider>
  )
})
