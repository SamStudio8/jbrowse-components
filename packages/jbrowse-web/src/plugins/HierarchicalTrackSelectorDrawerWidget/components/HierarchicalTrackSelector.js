import { IconButton } from '@material-ui/core'
import Checkbox from '@material-ui/core/Checkbox'
import ExpansionPanel from '@material-ui/core/ExpansionPanel'
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormGroup from '@material-ui/core/FormGroup'
import Icon from '@material-ui/core/Icon'
import InputAdornment from '@material-ui/core/InputAdornment'
import { withStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'
import Tooltip from '@material-ui/core/Tooltip'
import Typography from '@material-ui/core/Typography'
import { values } from 'mobx'
import { inject, observer, PropTypes as MobxPropTypes } from 'mobx-react'
import propTypes from 'prop-types'
import React from 'react'

const styles = theme => ({
  root: {
    textAlign: 'left',
    padding: theme.spacing.unit,
  },
  expansionPanelDetails: {
    display: 'block',
  },
})

function addTrackToHierarchy(trackHierarchy, track) {
  const numCategories = track.configuration.category.value.length
  if (numCategories === 0)
    trackHierarchy.set(
      'uncategorized',
      trackHierarchy.get('uncategorized').concat([track]),
    )
  else {
    let prunedHierarchy = trackHierarchy
    track.configuration.category.value.forEach((category, idx) => {
      if (!prunedHierarchy.has(category)) {
        prunedHierarchy.set(category, new Map([['uncategorized', []]]))
      }
      prunedHierarchy = prunedHierarchy.get(category)
      if (idx + 1 === numCategories) {
        prunedHierarchy.set(
          'uncategorized',
          prunedHierarchy.get('uncategorized').concat([track]),
        )
      }
    })
  }
}

function trackFilter(track, model) {
  return track.name.toLowerCase().includes(model.filterText)
}

function generateTrackList(trackHierarchy, view, classes, model) {
  const trackList = trackHierarchy.get('uncategorized')
  const elements = [
    <FormGroup key={`FormGroup-${trackList.map(track => track.id).join('-')}`}>
      {trackList
        .filter(track => trackFilter(track, model))
        .map(track => (
          <Tooltip
            key={track.id}
            title={track.configuration.description.value}
            placement="left"
            enterDelay={500}
          >
            <FormControlLabel
              key={track.id}
              control={<Checkbox />}
              label={track.name}
              checked={track.visible}
              onChange={() => {
                view.tracks.get(track.id).toggle()
              }}
            />
          </Tooltip>
        ))}
    </FormGroup>,
  ]
  trackHierarchy.forEach((value, category) => {
    if (category !== 'uncategorized') {
      /* eslint-disable react/no-array-index-key */
      elements.push(
        <ExpansionPanel
          key={category}
          expanded={model.categories.get(category).open}
          onChange={() => model.categories.get(category).toggle()}
        >
          <ExpansionPanelSummary
            key={category}
            expandIcon={<Icon>expand_more</Icon>}
          >
            <Typography key={category} variant="button">
              {`${category} (${
                value
                  .get('uncategorized')
                  .filter(track => trackFilter(track, model)).length
              })`}
            </Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails
            key={category}
            className={classes.expansionPanelDetails}
          >
            {generateTrackList(value, view, classes, model)}
          </ExpansionPanelDetails>
        </ExpansionPanel>,
      )
      /* eslint-enable react/no-array-index-key */
    }
  })
  return elements
}

@withStyles(styles)
@inject('rootModel')
@observer
class HierarchicalTrackSelector extends React.Component {
  static propTypes = {
    classes: propTypes.shape({
      root: propTypes.string.isRequired,
      expansionPanelDetails: propTypes.string.isRequired,
    }).isRequired,
    model: MobxPropTypes.observableObject.isRequired,
  }

  static wrappedComponent = {
    propTypes: {
      rootModel: MobxPropTypes.observableObject.isRequired,
    },
  }

  handleInputChange = event => {
    const { model } = this.props
    model.updateFilterText(event.target.value)
  }

  render() {
    const { classes, model, rootModel } = this.props
    const view = rootModel.views.filter(v => v.id === model.id)[0]
    const trackHierarchy = new Map([['uncategorized', []]])

    values(view.tracks).forEach(track => {
      addTrackToHierarchy(trackHierarchy, track)
    })

    const filterError =
      values(view.tracks).filter(track => trackFilter(track, model)).length ===
      0

    return (
      <div className={classes.root}>
        <TextField
          label="Filter Tracks"
          value={model.filterText}
          error={filterError}
          helperText={filterError ? 'No matches' : ''}
          onChange={this.handleInputChange}
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <Icon>search</Icon>
              </InputAdornment>
            ),
            endAdornment: (
              <InputAdornment position="start">
                <IconButton onClick={() => model.updateFilterText('')}>
                  <Icon>clear</Icon>
                </IconButton>
              </InputAdornment>
            ),
          }}
        />
        {generateTrackList(trackHierarchy, view, classes, model)}
      </div>
    )
  }
}

export default HierarchicalTrackSelector
