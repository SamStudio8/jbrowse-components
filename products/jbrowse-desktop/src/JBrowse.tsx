import React, { useEffect, useState, Suspense } from 'react'
import { observer } from 'mobx-react'
import { CssBaseline, ThemeProvider } from '@material-ui/core'
import { getConf } from '@jbrowse/core/configuration'
import { App, createJBrowseTheme } from '@jbrowse/core/ui'
import PluginManager from '@jbrowse/core/PluginManager'
import { AssemblyManager } from '@jbrowse/plugin-data-management'
import deepmerge from 'deepmerge'

// locals
import { RootModel } from './rootModel'

const JBrowse = observer(
  ({ pluginManager, mode }: { mode: string; pluginManager: PluginManager }) => {
    const { rootModel } = pluginManager

    return rootModel ? (
      <JBrowseNonNullRoot mode={mode} rootModel={rootModel as RootModel} />
    ) : null
  },
)

const JBrowseNonNullRoot = observer(
  ({ rootModel, mode }: { mode: string; rootModel: RootModel }) => {
    const [firstLoad, setFirstLoad] = useState(true)
    const { session, jbrowse, error, isAssemblyEditing, setAssemblyEditing } =
      rootModel

    useEffect(() => {
      if (firstLoad && session) {
        setFirstLoad(false)
      }
    }, [firstLoad, session])

    if (error) {
      throw error
    }

    const theme = getConf(jbrowse, 'theme')

    const colorTheme = React.useMemo(
      () => ({
        palette: {
          type: mode,
        },
      }),
      [mode],
    )

    return (
      // @ts-ignore
      <ThemeProvider theme={createJBrowseTheme(deepmerge(theme, colorTheme))}>
        <CssBaseline />
        {session ? (
          <>
            <App session={session} />
            <Suspense fallback={<div />}>
              {isAssemblyEditing ? (
                <AssemblyManager
                  rootModel={rootModel}
                  onClose={() => {
                    setAssemblyEditing(false)
                  }}
                />
              ) : null}
            </Suspense>
          </>
        ) : (
          <div>No session</div>
        )}
      </ThemeProvider>
    )
  },
)

export default JBrowse
