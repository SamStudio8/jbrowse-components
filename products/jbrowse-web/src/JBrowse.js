import { App, theme } from '@gmod/jbrowse-core/ui'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/core/styles'
import { observer } from 'mobx-react'
import { onSnapshot } from 'mobx-state-tree'
import { StringParam, useQueryParam } from 'use-query-params'
import React, { useEffect } from 'react'

// adapted from https://github.com/jashkenas/underscore/blob/5d8ab5e37c9724f6f1181c5f95d0020815e4cb77/underscore.js#L894-L925
function debounce(func, wait) {
  let timeout
  let result
  const later = (...args) => {
    timeout = null
    result = func(...args)
  }
  const debounced = (...args) => {
    if (timeout) {
      clearTimeout(timeout)
    }
    timeout = setTimeout(() => {
      return later(...args)
    }, wait)
    return result
  }
  debounced.cancel = () => {
    clearTimeout(timeout)
    timeout = null
  }
  return debounced
}

const JBrowse = observer(({ pluginManager }) => {
  const [sessionId] = useQueryParam('session', StringParam)

  const { rootModel } = pluginManager
  const { session, error } = rootModel || {}

  useEffect(() => {
    function updateLocalSession(snapshot) {
      // TODOSESSION have warning when reaching full local storage, have them manually delete instead of auto delete
      // sessionstorage over local storage, make save button that just transfers from session to local
      // autosave one in localstorage called local-autosaved that continuously updates on the most recent change
      console.log('updating-local-storage', sessionId)
      if (sessionId?.startsWith('local-'))
        localStorage.setItem(sessionId, JSON.stringify(snapshot))
    }

    let disposer = () => {}
    if (session) {
      const updater = debounce(updateLocalSession, 400)
      const snapshotDisposer = onSnapshot(session, updater)
      disposer = () => {
        snapshotDisposer()
        updater.cancel()
      }
    }
    return disposer
  }, [session, sessionId])

  if (error) {
    throw new Error(error)
  }

  return <App session={rootModel.session} />
})

export default props => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <JBrowse {...props} />
    </ThemeProvider>
  )
}
