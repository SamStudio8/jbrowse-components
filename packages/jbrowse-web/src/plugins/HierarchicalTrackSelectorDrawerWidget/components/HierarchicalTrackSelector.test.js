import React from 'react'
import { render } from 'react-testing-library'
import { createTestEnv } from '../../../JBrowse'
import HierarchicalTrackSelector from './HierarchicalTrackSelector'

describe('HierarchicalTrackSelector drawer widget', () => {
  it('renders with just the required model elements', async () => {
    const { rootModel } = await createTestEnv()
    const firstView = rootModel.addView('LinearGenomeView')
    firstView.activateTrackSelector()
    const model = rootModel.drawerWidgets.get('hierarchicalTrackSelector')

    const { container } = render(<HierarchicalTrackSelector model={model} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with a couple of uncategorized tracks', async () => {
    const { rootModel } = await createTestEnv({
      assemblies: {
        volvox: {
          tracks: [
            { configId: 'fooC', type: 'AlignmentsTrack' },
            { configId: 'barC', type: 'AlignmentsTrack' },
          ],
        },
      },
    })
    const firstView = rootModel.addView('LinearGenomeView')
    firstView.showTrack(
      rootModel.configuration.assemblies.get('volvox').tracks[0],
    )
    firstView.showTrack(
      rootModel.configuration.assemblies.get('volvox').tracks[1],
    )
    firstView.activateTrackSelector()
    const model = rootModel.drawerWidgets.get('hierarchicalTrackSelector')

    const { container } = render(<HierarchicalTrackSelector model={model} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with a couple of categorized tracks', async () => {
    const { rootModel } = await createTestEnv({
      assemblies: {
        volvox: {
          tracks: [
            { configId: 'fooC', type: 'AlignmentsTrack' },
            { configId: 'barC', type: 'AlignmentsTrack' },
          ],
        },
      },
    })
    const firstView = rootModel.addView('LinearGenomeView')
    firstView.showTrack(
      rootModel.configuration.assemblies.get('volvox').tracks[0],
    )
    firstView.showTrack(
      rootModel.configuration.assemblies.get('volvox').tracks[1],
    )
    firstView.tracks[0].configuration.category.set(['Foo Category'])
    firstView.tracks[1].configuration.category.set([
      'Foo Category',
      'Bar Category',
    ])
    firstView.activateTrackSelector()
    const model = rootModel.drawerWidgets.get('hierarchicalTrackSelector')

    const { container } = render(<HierarchicalTrackSelector model={model} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
