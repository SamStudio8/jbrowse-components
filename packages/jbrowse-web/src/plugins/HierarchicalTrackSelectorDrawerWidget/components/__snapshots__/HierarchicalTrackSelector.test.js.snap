// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchicalTrackSelector drawer widget renders with a couple of categorized tracks 1`] = `
<div
  class="HierarchicalTrackSelector-root-1"
>
  <div
    class="MuiFormControl-root-5 MuiFormControl-fullWidth-8 HierarchicalTrackSelector-searchBox-2"
  >
    <label
      class="MuiFormLabel-root-20 MuiFormLabel-error-23 MuiInputLabel-error-12 MuiInputLabel-root-9 MuiInputLabel-formControl-14 MuiInputLabel-animated-17 MuiInputLabel-shrink-16"
      data-shrink="true"
    >
      Filter Tracks
    </label>
    <div
      class="MuiInputBase-root-40 MuiInput-root-27 MuiInput-underline-31 MuiInputBase-error-46 MuiInput-error-32 MuiInputBase-fullWidth-49 MuiInput-fullWidth-34 MuiInputBase-formControl-41 MuiInput-formControl-28 MuiInputBase-adornedStart-44 MuiInputBase-adornedEnd-45"
    >
      <div
        class="MuiInputAdornment-root-57 MuiInputAdornment-positionStart-59"
      >
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root-62"
        >
          search
        </span>
      </div>
      <input
        aria-invalid="true"
        class="MuiInputBase-input-50 MuiInput-input-35 MuiInputBase-inputAdornedStart-55 MuiInputBase-inputAdornedEnd-56"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root-57 MuiInputAdornment-positionEnd-60"
      >
        <button
          class="MuiButtonBase-root-77 MuiIconButton-root-71"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-76"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root-62"
            >
              clear
            </span>
          </span>
          <span
            class="MuiTouchRipple-root-110"
          />
        </button>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root-80 MuiFormHelperText-error-81"
    >
      No matches
    </p>
  </div>
  <div
    class="MuiFormGroup-root-89"
  >
    <div
      class="TrackEntry-track-119"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <label
        class="MuiFormControlLabel-root-129 TrackEntry-formControlLabel-117"
        title="The reference sequence for volvox"
      >
        <span
          class="MuiButtonBase-root-77 MuiIconButton-root-71 MuiPrivateSwitchBase-root-141 MuiCheckbox-root-135 MuiCheckbox-colorSecondary-140 TrackEntry-checkbox-118"
        >
          <span
            class="MuiIconButton-label-76"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-145"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <input
              class="MuiPrivateSwitchBase-input-144"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
          </span>
          <span
            class="MuiTouchRipple-root-110"
          />
        </span>
        <span
          class="MuiTypography-root-154 MuiTypography-body1-163 MuiFormControlLabel-label-134"
        >
          Reference Sequence (volvox)
        </span>
      </label>
      <button
        class="MuiButtonBase-root-77 MuiIconButton-root-71 TrackEntry-configureButton-120"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-76"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root-62 MuiIcon-fontSizeSmall-69"
          >
            settings
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-110"
        />
      </button>
    </div>
  </div>
  <hr
    class="MuiDivider-root-91 Contents-divider-88"
  />
  <div
    class="MuiFormGroup-root-89"
  />
  <button
    class="MuiButtonBase-root-77 MuiFab-root-96 MuiFab-secondary-99 HierarchicalTrackSelector-fab-3"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label-97"
    >
      <span
        aria-hidden="true"
        class="material-icons MuiIcon-root-62"
      >
        add
      </span>
    </span>
    <span
      class="MuiTouchRipple-root-110"
    />
  </button>
</div>
`;

exports[`HierarchicalTrackSelector drawer widget renders with a couple of uncategorized tracks 1`] = `
<div
  class="HierarchicalTrackSelector-root-1"
>
  <div
    class="MuiFormControl-root-5 MuiFormControl-fullWidth-8 HierarchicalTrackSelector-searchBox-2"
  >
    <label
      class="MuiFormLabel-root-20 MuiFormLabel-error-23 MuiInputLabel-error-12 MuiInputLabel-root-9 MuiInputLabel-formControl-14 MuiInputLabel-animated-17 MuiInputLabel-shrink-16"
      data-shrink="true"
    >
      Filter Tracks
    </label>
    <div
      class="MuiInputBase-root-40 MuiInput-root-27 MuiInput-underline-31 MuiInputBase-error-46 MuiInput-error-32 MuiInputBase-fullWidth-49 MuiInput-fullWidth-34 MuiInputBase-formControl-41 MuiInput-formControl-28 MuiInputBase-adornedStart-44 MuiInputBase-adornedEnd-45"
    >
      <div
        class="MuiInputAdornment-root-57 MuiInputAdornment-positionStart-59"
      >
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root-62"
        >
          search
        </span>
      </div>
      <input
        aria-invalid="true"
        class="MuiInputBase-input-50 MuiInput-input-35 MuiInputBase-inputAdornedStart-55 MuiInputBase-inputAdornedEnd-56"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root-57 MuiInputAdornment-positionEnd-60"
      >
        <button
          class="MuiButtonBase-root-77 MuiIconButton-root-71"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-76"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root-62"
            >
              clear
            </span>
          </span>
          <span
            class="MuiTouchRipple-root-110"
          />
        </button>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root-80 MuiFormHelperText-error-81"
    >
      No matches
    </p>
  </div>
  <div
    class="MuiFormGroup-root-89"
  >
    <div
      class="TrackEntry-track-119"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <label
        class="MuiFormControlLabel-root-129 TrackEntry-formControlLabel-117"
        title="The reference sequence for volvox"
      >
        <span
          class="MuiButtonBase-root-77 MuiIconButton-root-71 MuiPrivateSwitchBase-root-141 MuiCheckbox-root-135 MuiCheckbox-colorSecondary-140 TrackEntry-checkbox-118"
        >
          <span
            class="MuiIconButton-label-76"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-145"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <input
              class="MuiPrivateSwitchBase-input-144"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
          </span>
          <span
            class="MuiTouchRipple-root-110"
          />
        </span>
        <span
          class="MuiTypography-root-154 MuiTypography-body1-163 MuiFormControlLabel-label-134"
        >
          Reference Sequence (volvox)
        </span>
      </label>
      <button
        class="MuiButtonBase-root-77 MuiIconButton-root-71 TrackEntry-configureButton-120"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-76"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root-62 MuiIcon-fontSizeSmall-69"
          >
            settings
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-110"
        />
      </button>
    </div>
  </div>
  <hr
    class="MuiDivider-root-91 Contents-divider-88"
  />
  <div
    class="MuiFormGroup-root-89"
  />
  <button
    class="MuiButtonBase-root-77 MuiFab-root-96 MuiFab-secondary-99 HierarchicalTrackSelector-fab-3"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label-97"
    >
      <span
        aria-hidden="true"
        class="material-icons MuiIcon-root-62"
      >
        add
      </span>
    </span>
    <span
      class="MuiTouchRipple-root-110"
    />
  </button>
</div>
`;

exports[`HierarchicalTrackSelector drawer widget renders with just the required model elements 1`] = `
<div
  class="HierarchicalTrackSelector-root-1"
>
  <div
    class="MuiFormControl-root-5 MuiFormControl-fullWidth-8 HierarchicalTrackSelector-searchBox-2"
  >
    <label
      class="MuiFormLabel-root-20 MuiFormLabel-error-23 MuiInputLabel-error-12 MuiInputLabel-root-9 MuiInputLabel-formControl-14 MuiInputLabel-animated-17 MuiInputLabel-shrink-16"
      data-shrink="true"
    >
      Filter Tracks
    </label>
    <div
      class="MuiInputBase-root-40 MuiInput-root-27 MuiInput-underline-31 MuiInputBase-error-46 MuiInput-error-32 MuiInputBase-fullWidth-49 MuiInput-fullWidth-34 MuiInputBase-formControl-41 MuiInput-formControl-28 MuiInputBase-adornedStart-44 MuiInputBase-adornedEnd-45"
    >
      <div
        class="MuiInputAdornment-root-57 MuiInputAdornment-positionStart-59"
      >
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root-62"
        >
          search
        </span>
      </div>
      <input
        aria-invalid="true"
        class="MuiInputBase-input-50 MuiInput-input-35 MuiInputBase-inputAdornedStart-55 MuiInputBase-inputAdornedEnd-56"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root-57 MuiInputAdornment-positionEnd-60"
      >
        <button
          class="MuiButtonBase-root-77 MuiIconButton-root-71"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-76"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root-62"
            >
              clear
            </span>
          </span>
          <span
            class="MuiTouchRipple-root-110"
          />
        </button>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root-80 MuiFormHelperText-error-81"
    >
      No matches
    </p>
  </div>
  <div
    class="MuiFormGroup-root-89"
  />
  <hr
    class="MuiDivider-root-91 Contents-divider-88"
  />
  <div
    class="MuiFormGroup-root-89"
  />
  <button
    class="MuiButtonBase-root-77 MuiFab-root-96 MuiFab-secondary-99 HierarchicalTrackSelector-fab-3"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label-97"
    >
      <span
        aria-hidden="true"
        class="material-icons MuiIcon-root-62"
      >
        add
      </span>
    </span>
    <span
      class="MuiTouchRipple-root-110"
    />
  </button>
</div>
`;
